service/raft/raft_group0.cc

```diff
@@ -141,6 +144,8 @@ class group0_rpc: public service::raft_rpc {
             _address_map.set_expiring(addr.id);
             _direct_fd.remove_endpoint(addr.id.id);
         }
+
+        _state_id_handler.on_configuration_change(add, del);
```

Comment:
Will this be called with the full configuration when the server is first started?

Reply:
Also, how does it interact with joint configurations? Is a raft server included in `add` when it first appears during joint configuration, or after we finish the config change? Symmetric question for `del`

Reply:
From the `server_impl::process_fsm_output()` it looks that the config is first changed (`add_to_rpc_config(addr)`) and then the `on_configuration_change()` is called with the added servers. In the case of removal, it looks again the the configuration is first changed (`remove_from_rpc_config(addr)`) and then the `on_configuration_change()` is called.

Reply:
That describes configuration changes while raft is running. The question is how does the handler learn about initial set of members that we know of e.g. on node restart

Reply:
Not really sure how it exactly works, but from is is called from the server startup code in here:
https://github.com/scylladb/scylladb/blob/master/raft/server.cc#L396-L407

```cpp
        // Account both for current and previous configurations since
        // the last configuration idx can point to the joint configuration entry.
        rpc_config.current.merge(rpc_config.previous);
        for (const auto& s: rpc_config.current) {
            add_to_rpc_config(s.addr);
        }
        _rpc->on_configuration_change(get_rpc_config(), {});
```

so again it looks to me like this is called with the whole configuration after it is merged together (?)

Reply:
Please double-check that I didn't slip up in my reasoning

Reply:
Yes, that sounds reasonable to me as well